### PR TITLE
Add output stream changing for created formatters

### PR DESCRIPTION
**Pull request by dzjuck**
Reason of bug described in #2216. I think easiest way to fix - update `output` of already created formatters at every change of config `output_stream`.

I’m not sure how tests should be written for this fix. So if any thoughts how to tests should be done or how to make solution better - i will glad to make it.

### CODE REVIEW

File: lib/rspec/core/formatters/base_formatter.rb [hunk 1]
```diff
@@ -51,6 +51,11 @@ def close(_notification)
           restore_sync_output
         end
 
+        # @private
+        def change_output(output)
```
Comment (JonRowe) [inline lib/rspec/core/formatters/base_formatter.rb]: Our convention would just be to create an `attr_writer` or `attr_accessor` for output, it also can't be private as every formatter will have to implement this method.

File: spec/rspec/core/configuration_spec.rb [hunk 2]
```diff
@@ -125,6 +125,12 @@ module RSpec::Core
           expect(config).not_to have_received(:warn)
         end
       end
+
+      it 'changes output stream of already created formatters' do
+        io = double 'output io'
+        expect(config.formatter_loader).to receive(:change_output_stream).with(io)
+        config.output_stream = io
```
Comment (JonRowe) [inline spec/rspec/core/configuration_spec.rb]: We use a `StringIO` to represent IO objects in this spec, would be nice to match that convention here, it has the advantage of being a ready made "double" for IO.

File: spec/rspec/core/formatters_spec.rb [hunk 3]
```diff
@@ -206,5 +206,14 @@ module RSpec::Core::Formatters
         end
       end
     end
+
+    describe "#change_output_stream" do
+      let(:new_output) { double 'output io' }
+      it "change output stream to all formatters" do
+        loader.add :documentation, output
+        expect(loader.formatters.first).to receive(:change_output).with(new_output)
+        loader.change_output_stream(new_output)
```
Comment (JonRowe) [inline spec/rspec/core/formatters_spec.rb]: I think it'd be nicer to use a `StringIO` here and assert that future notifications write into it.

File: lib/rspec/core/formatters.rb [hunk 4]
```diff
@@ -170,6 +170,11 @@ def add(formatter_to_use, *paths)
       end
     end
 
+    # @private
+    def change_output_stream(output_stream)
+      @formatters.each { |formatter| formatter.change_output(output_stream) }
```
Comment (JonRowe) [inline lib/rspec/core/formatters.rb]: This needs to cope with formatters not implementing this method until RSpec 4, as it's a breaking change in our API. Formatters can be defined in many ways not just inheriting from our classes.

File: features/configuration/output_stream.feature [hunk 5]
```diff
@@ -23,5 +23,5 @@ Feature: Custom output stream
         end
       end
       """
-    When I run `rspec spec/example_spec.rb`
+    When I run `rspec spec/example_spec.rb --format progress`
```
Comment (JonRowe) [inline features/configuration/output_stream.feature]: This is unrelated as far as I can tell and should be reverted, if it's not unrelated please explain why it's here and it'd be better to have a whole scenario for it.

File: spec/rspec/core/configuration_spec.rb [hunk 6]
```diff
@@ -125,6 +125,12 @@ module RSpec::Core
           expect(config).not_to have_received(:warn)
         end
       end
+
+      it 'changes output stream of already created formatters' do
```
Comment (JonRowe) [inline spec/rspec/core/configuration_spec.rb]: Very minior but "changes the output stream"

File: spec/rspec/core/formatters_spec.rb [hunk 7]
```diff
@@ -206,5 +206,14 @@ module RSpec::Core::Formatters
         end
       end
     end
+
+    describe "#change_output_stream" do
+      let(:new_output) { double 'output io' }
+      it "change output stream to all formatters" do
```
Comment (JonRowe) [inline spec/rspec/core/formatters_spec.rb]: s/change/changes the/

File: lib/rspec/core/formatters.rb (hunk 4 above)
Comment (myronmarston) [inline lib/rspec/core/formatters.rb]: Given that this only works for our formatters and ones that inherit ours, it makes me wonder if a different way of doing this would be better.  One idea is to not pass the formatters a raw `output` object but instead a wrapper object that holds a reference to a real `output` internally.  If the `formatter_loader` also has a reference to the same `output_wrapper` object that each formatter is given, the `formatter_loader` can tell the `output_wrapper` to change where it outputs to, and all the formatters should "just work" with no changes needed.  Thoughts?

Comment (JonRowe) [inline lib/rspec/core/formatters.rb]: Well I guess we'd only be mutating state in one place then, seems sensible to me!

**Comment by dzjuck** [comment]
@JonRowe What kind of integration spec should do? I see cucumber’s `features` and `spec/integration`.

File: lib/rspec/core/formatters.rb (hunk 4 above)
Comment (dzjuck) [inline lib/rspec/core/formatters.rb]: Great. I will do `output_wrapper`.

**Comment by myronmarston** [comment]
> @JonRowe What kind of integration spec should do? I see cucumber’s features and spec/integration.

Please use `spec/integration`.  We try to limit cucumber features to things that have documentation value, but this would just demonstrate "that config setting we documented elsewhere also works in this edge case"--which is just noise in docs, I think.

**Comment by dzjuck** [comment]
@myronmarston Okay, i will make test in `spec/integration`

**Comment by dzjuck** [comment]
It’s temporary commit. I want to clarify `output_wrapper` realisation.

I made `OutputWrapper` as subclass of `SimpleDelegator`. Everything works fine except one thing. Formatters behaviour may depend on `output` class. For example `base_text_formatter` contains a [condition](https://github.com/rspec/rspec-core/blob/master/lib/rspec/core/formatters/base_text_formatter.rb#L66) `return unless IO === output`. I don’t see way to make it works for `OutputWrapper`.

It’s possible to override method `OutputWrapper#class` and rewrite the condition as `return unless output.class == IO`. But it may breaks custom formatters.

Any thought how to solve this problem?

File: lib/rspec/core/output_wrapper.rb [hunk 8]
```diff
@@ -0,0 +1,11 @@
+require 'delegate'
+
+class OutputWrapper < SimpleDelegator
```
Comment (JonRowe) [inline lib/rspec/core/output_wrapper.rb]: Please don't use SimpleDelgegator, it's easy to reimplement and we believe in using as little of the standard library as possible to avoid poisoning a users environment. SimpleDelgator also has performance issues on certain versions of Ruby.

Instead implement the IO public protocol methods and directly delegate to output.

**Comment by JonRowe** [comment]
We could use IO as the base class, then override all the methods... then it would be `===` etc to `IO`

**Comment by dzjuck** [comment]
@JonRowe This should work. Thanks.

File: lib/rspec/core/output_wrapper.rb (hunk 8 above)
Comment (dzjuck) [inline lib/rspec/core/output_wrapper.rb]: IO class adds 70 public methods. Should `output_wrapper` implement all of them? RSpec code uses only 6 methods of IO `tty? sync= closed? puts print flush`.

Also i suppose for perfomance better to implement methods without using of meta programming, like this:

```ruby
  def puts(*args)
    @output.puts(args)
  end

```

**Comment by JonRowe** [comment]
It's  be nice to, you can do this programmatically by looping over an array
with define_method

**Comment by dzjuck** [comment]
`OutputWrapper` as `IO` subclass does work for `StriongIO` class (because `StriongIO` is not `IO` subclass).

```ruby
Wrapper = OutputWrapper.new($stdout)
IO === wrapper # should be true

Wrapper.output = StringIO.new
IO === wrapper # should be false
```
 
Also i suppose reason behind `IO === output` condition is to separate a behaviour for StringIO and File/IO.

**Comment by myronmarston** [comment]
You know, I don't think we need to maintain `IO === output`.  I'm not sure why we have that conditional.  I looked at the git history, and that line goes all the way back to micronaut (the runner that became rspec-core 2):

https://github.com/spicycode/micronaut/commit/b581fc176f397bfeb5716990147043794bf48f4c#diff-414c0b2d7d897134aacf8c77c43bbe7cR77

It must have been needed at that point, but at this point I don't think we need it.  I think we can just remove the `return unless IO === output` line entirely.  Note that removing it does make some specs fail, but the failures actually illustrate an issue that line was creating: it was making the output be slightly different for a `StringIO` vs a real IO like `$stdout`.  With a real IO, an extra newline was being printed, but not for `StringIO`, and many tests use `StringIO` and have expectations about the output that now fail due to lacking the extra newline.  I don't think we want `StringIO` to get different output than a real IO--in a test, you use `StringIO` to be able to make assertions about what the output would be in a real situation, so it being different is a bad thing.  If you remove it you'll need to update those failing specs.

**Comment by dzjuck** [comment]
It makes sense. I removed `return unless IO === output` and fixed specs. And ready to improve the solution.

File: lib/rspec/core/configuration.rb [hunk 9]
```diff
@@ -870,7 +874,7 @@ def full_description
       # Adds a formatter to the set RSpec will use for this run.
       #
       # @see RSpec::Core::Formatters::Protocol
-      def add_formatter(formatter, output=output_stream)
+      def add_formatter(formatter, output=output_wrapper)
```
Comment (JonRowe) [inline lib/rspec/core/configuration.rb]: I think we should take the given output stream and wrap it, the way this is currently it will allow someone to not use a wrapper and thus re trigger the bug

File: lib/rspec/core/output_wrapper.rb [hunk 10]
```diff
@@ -0,0 +1,26 @@
+module RSpec
+  module Core
+    # Wrapper around an output stream object.
```
Comment (JonRowe) [inline lib/rspec/core/output_wrapper.rb]: This whole class should be private API IMO

File: lib/rspec/core/output_wrapper.rb [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    class OutputWrapper
+      # @private
+      attr_accessor :output
```
Comment (myronmarston) [inline lib/rspec/core/output_wrapper.rb]: Given you use the instance variable directly and don't use the reader, should this be just `attr_writer`?

File: spec/rspec/core/output_wrapper_spec.rb [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+        it "calls output's '#{method}' with the same params" do
+          expect(output).to receive(method).with('all', 'args')
+          wrapper.send(method, 'all', 'args')
+        end
+        it "calls output's '#{method}' with the same params and block" do
+          expect(output).to receive(method).with('all', 'args', &block)
+          wrapper.send(method, 'all', 'args', &block)
+        end
+      end
+    end
```
Comment (myronmarston) [inline spec/rspec/core/output_wrapper_spec.rb]: I'm not sure that testing each method like this really does anything for us other than bloating our spec site with tons of specs.  Given they are defined in a loop I'm not too worried about a particular one not working properly. Plus this is going to be exercised constantly by all the other specs (anything involving formatter output!) that I'm not worried about it breaking on us.  

Instead, I think I'd like to see specs that demonstrate what this class is for: acting like an IO while allowing redirection to a different IO after instantiation.  To that end, I think you can write some a spec that instantiates it with a `StringIO` and calls IO methods to show that it writes to the underlying StringIO.  Then you can write another spec that shows that when you change the output it writes new output to that StringIO instance instead.

File: lib/rspec/core/output_wrapper.rb [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+        @output = output
+      end
+
+      [:<<, :advise, :autoclose=, :autoclose?, :binmode, :binmode?, :bytes, :chars, :close, :close_on_exec=,
+       :close_on_exec?, :close_read, :close_write, :closed?, :codepoints, :each, :each_byte, :each_char,
+       :each_codepoint, :each_line, :eof, :eof?, :external_encoding, :fcntl, :fdatasync, :fileno, :flush,
+       :fsync, :getbyte, :getc, :gets, :inspect, :internal_encoding, :ioctl, :isatty, :lineno, :lineno=,
+       :lines, :pid, :pos, :pos=, :print, :printf, :putc, :puts, :read, :read_nonblock, :readbyte, :readchar,
+       :readline, :readlines, :readpartial, :reopen, :rewind, :seek, :set_encoding, :stat, :sync, :sync=,
+       :sysread, :sysseek, :syswrite, :tell, :to_i, :to_io, :tty?, :ungetbyte, :ungetc, :write, :write_nonblock].each do |method|
```
Comment (myronmarston) [inline lib/rspec/core/output_wrapper.rb]: Can you add a comment explanining how you came up with this list of methods?

Comment (dzjuck) [inline lib/rspec/core/output_wrapper.rb]: Sure, i will add a comment to the code. Also this is list of `IO` methods for Ruby >= 1.9.3. And Ruby 1.8.7 `IO` has different set of methods and they are not subset of list above. I noticed Ruby 1.8.7 is currently support, so should i make different methods lists for Ruby >= 1.9.3 and Ruby 1.8.7?

Comment (myronmarston) [inline lib/rspec/core/output_wrapper.rb]: Having completely separate lists sounds painful, but I think this could work:

``` ruby
io_methods = [ # full list here ]
if RUBY_VERSION < '1.9'
  io_methods -= [ # methods only available on 1.9+ ]
end

io_methods.each do |method|
  # define the method
end
```

Comment (JonRowe) [inline lib/rspec/core/output_wrapper.rb]: Why don't we just use`IO.instance_methods` to generate the list at runtime?

Comment (dzjuck) [inline lib/rspec/core/output_wrapper.rb]: Good idea. I suppose we want to decorate only IO specific methods, not Object methods. I just have tried, `IO.instance_methods - Enumerable.instance_methods - Object.instance_methods` gives same list as above except `:inspect` method. Specs are green.

File: lib/rspec/core/configuration.rb (hunk 9 above)
Comment (dzjuck) [inline lib/rspec/core/configuration.rb]: It’s possible to send specific output stream to `add_formatter`:

output = StringIO.new
RSpec.configure do |c|
  c.add_formatter :documentation
  c.add_formatter :progress, output  
  c.output_stream = File.open('saved_output', 'w')
end

In this case `c.output_stream = File.open('saved_output', 'w')` should change output stream for `:documentation` formatter, but not for `:progress` formatter.

So i suppose,  it’s not obligatorily to wrap output param in add_formatter. Anyway maybe it worth to wrap output param here, in that case formatters always will operate with an object of `OutputWrapper`.

**Comment by dzjuck** [comment]
I updated pr. Pls check.

File: lib/rspec/core/configuration.rb (hunk 9 above)
Comment (JonRowe) [inline lib/rspec/core/configuration.rb]: This is why it's necessary to wrap the passed in input here, a user can add their own output and we need to wrap it.

**Comment by JonRowe** [comment]
This just needs to make sure we wrap a passed in users output in the wrapper and we're good to go

File: lib/rspec/core/configuration.rb (hunk 9 above)
Comment (dzjuck) [inline lib/rspec/core/configuration.rb]: Sounds reasonable, i've added wrapping

File: lib/rspec/core/output_wrapper.rb [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+      (IO.instance_methods - Enumerable.instance_methods - Object.instance_methods + [:to_str]).each do |method|
+        define_method(method) do |*args, &block|
+          @output.send(method, *args, &block)
+        end
+      end
+
+      # @private
+      def respond_to?(*args)
+        args[0] == :output= || @output.respond_to?(*args)
+      end
```
Comment (dzjuck) [inline lib/rspec/core/output_wrapper.rb]: Exists case when `output` will be an string with output file name.
In that case `wrapper` should not respond to `IO` methods and i added `respond_to?` method.
I'm not sure the result implementation is good enough.
May be better to write like this:
```ruby
METHODS = IO.instance_methods - Enumerable.instance_methods - Object.instance_methods

if METHODS.include?(args[0])
  @output.respond_to?(*args)
else
  super
end
```
I guess, the last code looks better, but performance will be little bit worse.

File: lib/rspec/core/output_wrapper.rb [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+      # @private
+      attr_writer :output
+
+      # @private
+      def initialize(output)
+        @output = output
+      end
+
+      # Redirect calls for IO interface methods
+      (IO.instance_methods - Enumerable.instance_methods - Object.instance_methods + [:to_str]).each do |method|
```
Comment (myronmarston) [inline lib/rspec/core/output_wrapper.rb]: Can you just use `IO.instance_methods(false)` instead of subtracting the `Enumerable` and `Object` methods?

Also, what's special about `to_str` that requires it to be added here?  It's not clear why it should be delegated...

File: lib/rspec/core/output_wrapper.rb (hunk 14 above)
Comment (myronmarston) [inline lib/rspec/core/output_wrapper.rb]: Given this class is designed to wrap an `IO` object, and a string is _not_ an `IO` object, I think we should not wrap a string with this class (and therefore, you shouldn't need this extra logic).  Instead, can the wrapping be delayed until the file identified by the string is opened?

Comment (myronmarston) [inline lib/rspec/core/output_wrapper.rb]: Actually, do we even support `output_stream` being set to a file name?  I can't find anything that would make that work...

File: lib/rspec/core/configuration.rb [hunk 16]
```diff
@@ -870,7 +874,8 @@ def full_description
...
+        output = OutputWrapper.new(output) unless output == output_wrapper
```
Comment (JonRowe) [inline lib/rspec/core/configuration.rb]: Is there a reason why this can't just set the output on the existing wrapper, and pass that through?

File: lib/rspec/core/output_wrapper.rb (hunk 14 above)
Comment (JonRowe) [inline lib/rspec/core/output_wrapper.rb]: I don't think we support that...

Comment (dzjuck) [inline lib/rspec/core/output_wrapper.rb]: I agree, the wrapping should be called after a string became an `IO` object.
`output_stream` may be a string when it sent as a command line param:
Rspec --format progress --out specs.out

A description of  `Configuration#add_formatter#870` says:
      # @param output [String, IO] where the formatter will write its output.
      #   Can be an IO object or a string path to a file. ...

`Formatter::Loader#add` [contains](https://github.com/rspec/rspec-core/blob/master/lib/rspec/core/formatters.rb#L148) the code to check the `output_stream` interface:

```ruby
args = paths.map { |p| p.respond_to?(:puts) ? p : file_at(p) }
```

And [creates](https://github.com/rspec/rspec-core/blob/master/lib/rspec/core/formatters.rb#L256) file when the `output_stream` is a file path:

```ruby
def file_at(path)
  RSpec::Support::DirectoryMaker.mkdir_p(File.dirname(path))
  File.new(path, 'w')
end
```

I.e. `--out specs.out` creates `File.new('specs.out', 'w')`.

Now RSpec::Core::Formatters::Loader#add has two responsibilities:

1. To create a file if a `path` is a string
2. To register a formatter

We may move out creation of a file from `Loader#add`. For example on the level of `Configuration#add_formatter` when `Loader#add` always receives an `IO` object and `OutputWrapper` wraps only the `IO` interface.

File: lib/rspec/core/configuration.rb (hunk 16 above)
Comment (dzjuck) [inline lib/rspec/core/configuration.rb]: Now the code of `Configuration#add_formatter` is:

```ruby
def add_formatter(formatter, output=output_stream)
  formatter_loader.add(formatter, output)
end
```

As you may see, `add_formatter` does change the `output_stream` in any way. If we set the `output` param to the existing wrapper, it is equivalent to the following code (in the version without wrapper):

```ruby
def add_formatter(formatter, output=output_stream)
  @output_stream = output
  formatter_loader.add(formatter, @output_stream)
end
```

It changes the current behaviour of `add_formatter` where you could couple any formatter with a specific `output`, which does not relate to the common `Configuration#output_stream`. In my opinion that behavior was quite logical.

Also i did not find any specific documentation or a test about this case. And if you change `add_formatter` to:

```ruby
def add_formatter(formatter, output=output_stream)
  @output_stream = output
  formatter_loader.add(formatter, @output_stream)
end
```

all specs are still green meaning it’s not covered by tests. 
But i’m not sure it is reasonable to change the current behaviour of `add_formatter` here.

Comment (JonRowe) [inline lib/rspec/core/configuration.rb]:
```
def add_formatter(formatter, output=output_wrapper.output)
  output_wrapper.output = output
  formatter_loader.add(formatter, output_wrapper)
end
```

Comment (dzjuck) [inline lib/rspec/core/configuration.rb]: Great! I will do it.

File: lib/rspec/core/configuration.rb [hunk 17]
```diff
@@ -226,7 +227,8 @@ def output_stream=(value)
             "`output_stream` will be ignored. You should configure it earlier for " \
```
Comment (myronmarston) [inline lib/rspec/core/configuration.rb]: Mind updating the yard doc on this method to document it can be a string since it apparently can be?

File: lib/rspec/core/configuration.rb [hunk 18]
```diff
@@ -908,6 +913,11 @@ def formatter_loader
       end
 
       # @private
+      def output_wrapper
```
Comment (myronmarston) [inline lib/rspec/core/configuration.rb]: Does this method need to be public?  While you've marked it private with a YARD annotation it's still public and it would be nice to make it private if possible..

File: lib/rspec/core/output_wrapper.rb [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+
+      # Redirect calls for IO interface methods
+      IO.instance_methods(false).each do |method|
+        define_method(method) do |*args, &block|
+          @output.send(method, *args, &block)
+        end
+      end
+    end
+  end
+end
```
Comment (myronmarston) [inline lib/rspec/core/output_wrapper.rb]: :+1: Nice to see how simple this is!

File: spec/integration/output_stream_spec.rb [hunk 20]
```diff
@@ -0,0 +1,30 @@
+require 'support/aruba_support'
+
+RSpec.describe 'Output stream' do
+  include_context 'aruba support'
+  before { clean_current_dir }
+
+  context 'when a formatter setted in a configure block' do
```
Comment (myronmarston) [inline spec/integration/output_stream_spec.rb]: s/setted/set/

"Set" is already past tense.  (English is weird).

File: lib/rspec/core/configuration.rb [hunk 21]
```diff
@@ -226,7 +227,8 @@ def output_stream=(value)
             "`output_stream` will be ignored. You should configure it earlier for " \
             "it to take effect. (Called from #{CallerFilter.first_non_rspec_line})"
         else
-          @output_stream = value
+          @output_stream = prepare_output(value)
```
Comment (JonRowe) [inline lib/rspec/core/configuration.rb]: This also needs to be used on outputs passed in via `add_formatter`

File: lib/rspec/core/configuration.rb [hunk 22]
```diff
@@ -471,6 +473,9 @@ def initialize
       # Used to set higher priority option values from the command line.
       def force(hash)
         ordering_manager.force(hash)
+        [:deprecation_stream, :output_stream].each do |stream|
+          hash[stream] = prepare_stream(hash[stream]) if hash.key?(stream)
+        end
```
Comment (dzjuck) [inline lib/rspec/core/configuration.rb]: There is one more place where `deprecation_stream` can be set.

File: features/command_line/format_option.feature [hunk 23]
```diff
@@ -72,15 +72,3 @@ Feature: `--format` option
...
-    When I run `rspec example_spec.rb --format progress --format documentation --out rspec.txt`
-    Then the output should contain ".F**"
-    And the file "rspec.txt" should contain:
-      """
-      something
-        does something that passes
-        does something that fails (FAILED - 1)
-        does something that is pending (PENDING: No reason given)
-        does something that is skipped (PENDING: No reason given)
-      """
```
Comment (dzjuck) [inline features/command_line/format_option.feature]: Now both formatters will be directed to `rspec.txt` and this will give different result on different Ruby realizations.

Comment (JonRowe) [inline features/command_line/format_option.feature]: Er... If that's true we need to fix it, the point of this feature is that the first formatter uses the default output, and the second uses the rspec.txt file, changing that is a breaking change.

Comment (dzjuck) [inline features/command_line/format_option.feature]: Ok, then i will restore the previous logic in `Configuration#add_formatter`

File: lib/rspec/core/configuration.rb [hunk 24]
```diff
@@ -471,6 +473,9 @@ def initialize
       # Used to set higher priority option values from the command line.
       def force(hash)
         ordering_manager.force(hash)
+        [:deprecation_stream, :output_stream].each do |stream|
+          hash[stream] = prepare_stream(hash[stream]) if hash.key?(stream)
```
Comment (myronmarston) [inline lib/rspec/core/configuration.rb]: This is mutating the passed in `hash`.  In general, it's a bad idea to mutate arguments passed to methods as it leads to unpleasant surprises.

File: lib/rspec/core/configuration.rb [hunk 25]
```diff
@@ -141,7 +142,7 @@ def expose_dsl_globally=(value)
 
       # Determines where deprecation warnings are printed.
       # Defaults to `$stderr`.
-      # @return [IO, String] IO to write to or filename to write to
+      # @return [IO] IO to write to
```
Comment (JonRowe) [inline lib/rspec/core/configuration.rb]: Do we not allow a string here, given it was documented it should...

File: lib/rspec/core/configuration.rb (hunk 24 above)
Comment (JonRowe) [inline lib/rspec/core/configuration.rb]: Yep if the intent is to pass a modified hash down to ordering_manager, use merge on the hash as it's passed in e.g.

```
ordering_manager.force(hash.merge(
...
))
```

Comment (dzjuck) [inline lib/rspec/core/configuration.rb]: Thank you. It’s my mistake, i did not intend to change the original hash. I will change the code.

File: lib/rspec/core/configuration.rb (hunk 25 above)
Comment (dzjuck) [inline lib/rspec/core/configuration.rb]: We moved stream preparation from `Formatters#Loader` to `Configuration`. And now `deprecation_stream` can’t contain a string here, because a deprecation stream already will be prepared and a string will be transformed to a file object.

**Comment by JonRowe** [comment]
Thanks for all your work on this @dzjuck I have some lingering concerns but we want to get this into 3.6 so I'm going to merge and then open up a follow up pr!

**Comment by dzjuck** [comment]
@JonRowe @myronmarston Thanks for your help and patience.
